lib/logstash/inputs/redis.rb

```diff
@@ -73,38 +94,79 @@ def register
       @data_type = 'list'
     end
 
-    if not @key or not @data_type
+    if !@key || !@data_type
       raise RuntimeError.new(
         "Must define queue, or key and data_type parameters"
       )
     end
     # end TODO
 
-    @logger.info("Registering Redis", :identity => identity)
+    @redis_builder ||= method(:internal_redis_builder)
+
+    # just switch on data_type once
+    if @data_type == 'list' || @data_type == 'dummy'
+      @run_method = method(:list_runner)
+      @teardown_method = method(:list_teardown)
+    elsif @data_type == 'channel'
+      @run_method = method(:channel_runner)
+      @teardown_method = method(:subscribe_teardown)
+    elsif @data_type == 'pattern_channel'
+      @run_method = method(:pattern_channel_runner)
+      @teardown_method = method(:subscribe_teardown)
+    end
+
+    # TODO(sissel, boertje): set @identity directly when @name config option is removed.
+    @identity = @name != 'default' ? @name : "#{@redis_url} #{@data_type}:#{@key}"
+    @logger.info("Registering Redis", :identity => @identity)
   end # def register
 
-  # A string used to identify a Redis instance in log messages
-  # TODO(sissel): Use instance variables for this once the @name config
-  # option is removed.
-  private
-  def identity
-    @name || "#{@redis_url} #{@data_type}:#{@key}"
+  def run(output_queue)
+    @run_method.call(output_queue)
+  rescue LogStash::ShutdownSignal
+    # ignore and quit
+  end # def run
+
+  def teardown
+    @shutdown_requested = true
+    @teardown_method.call
   end
 
+  # private methods -----------------------------
   private
-  def connect
-    redis = Redis.new(
+
+  def batched?
+    @batch_count > 1
+  end
+
+  # private
+  def is_list_type?
+    @data_type == 'list'
+  end
+
+  # private
+  def redis_params
+    {
       :host => @host,
       :port => @port,
       :timeout => @timeout,
       :db => @db,
       :password => @password.nil? ? nil : @password.value
-    )
-    load_batch_script(redis) if @data_type == 'list' && (@batch_count > 1)
-    return redis
+    }
+  end
+
+  # private
+  def internal_redis_builder
```

Comment:
I'm with Brandon here: http://brandonhilkert.com/blog/a-ruby-refactor-exploring-dependency-injection-options/

IMO, the Rubyist's aversion to DI is an embracing of the dynamic nature of Ruby and the reaction to the DI ceremony coming from Enterprise Java.

To use RSpec mocking without injecting a double into the Subject, one has to use the expect_any_instance_of or allow_any_instance_of mechanisms. The RSpec authors discourage its use, though: https://github.com/rspec/rspec-mocks#settings-mocks-or-stubs-on-any-instance-of-a-class.  I'm OK with doing this, but I would rather we debated this very subject as a team to settle on a acceptable pattern going forward.

RSpec 3 has object doubles where one can augment an existing instance. To do this when the unit-under-test builds its own collaborators from configs, one needs to have public methods that will allow one to: get a built collaborator (so the test does not need to know how to do this); augment that instance; inject it back into the UUT.

There are two debateable issues with DI, 1) DI to get an alternative collaborator object into the receiver for the purposes of testing and 2) simply getting an alternative collaborator (with the same public API, or subset of just what is needed) into the receiver at runtime.